lib/DependencyProvider.js

```diff
@@ -0,0 +1,367 @@
+// @flow
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const wasm = require('elm-solve-deps-wasm');
+const SyncGet = require('./SyncGet.js');
+const collator = new Intl.Collator('en', { numeric: true }); // for sorting SemVer strings
+
+// Initialization work done only once.
+wasm.init();
+const syncGetWorker /*: {| get: (string) => string, shutDown: () => void |} */ =
+  SyncGet.startWorker();
+
+// Cache of existing versions according to the package website.
+let onlineVersionsCache /*: Map<string, Array<string>> */ = new Map();
+
+// Memoization cache to avoid doing the same work twice in listAvailableVersions.
+// This is to be cleared before each call to solve_deps().
+const listVersionsMemoCache /*: Map<string, Array<string>> */ = new Map();
+
+// Constructor of an offline solver.
+function newOffline() /*: {
+  solve: (string, boolean, { [string]: string }) => string,
+} */ {
+  function solve(elmJson, useTest, extra) {
+    listVersionsMemoCache.clear();
+    try {
+      return wasm.solve_deps(
+        elmJson,
+        useTest,
+        extra,
+        fetchElmJsonOffline,
+        listAvailableVersionsOffline
+      );
+    } catch (errorMessage) {
+      throw new Error(errorMessage);
+    }
+  }
+  return { solve };
+}
+
+// Constructor of an online solver.
+function newOnline() /*: {
+  solve: (string, boolean, { [string]: string }) => string,
+} */ {
+  function solve(elmJson, useTest, extra) {
+    updateOnlineVersionsCache();
+    listVersionsMemoCache.clear();
+    try {
+      return wasm.solve_deps(
+        elmJson,
+        useTest,
+        extra,
+        fetchElmJsonOnline,
+        listAvailableVersionsOnline
+      );
+    } catch (errorMessage) {
+      throw new Error(errorMessage);
+    }
+  }
+  return { solve };
+}
+
+function fetchElmJsonOnline(
+  pkg /*: string */,
+  version /*: string */
+) /*: string */ {
+  try {
+    return fetchElmJsonOffline(pkg, version);
+  } catch (_) {
+    // `fetchElmJsonOffline` can only fail in ways that are either expected
+    // (such as file does not exist or no permissions)
+    // or because there was an error parsing `pkg` and `version`.
+    // In such case, this will throw again with `cacheElmJsonPath()` so it's fine.
+    const remoteUrl = remoteElmJsonUrl(pkg, version);
+    const elmJson = syncGetWorker.get(remoteUrl);
+    const cachePath = cacheElmJsonPath(pkg, version);
+    const parentDir = path.dirname(cachePath);
+    fs.mkdirSync(parentDir, { recursive: true });
+    fs.writeFileSync(cachePath, elmJson);
+    return elmJson;
+  }
+}
+
+function fetchElmJsonOffline(
+  pkg /*: string */,
+  version /*: string */
+) /*: string */ {
+  // console.log('Fetching: ' + pkg + ' @ ' + version);
```

Comment:
Can we remove all logging leftovers now? (Search for `// console`)

Reply:
Ah I forgot about those. I wanted to ask if there is a logging mechanism in elm-test which can be enabled with configurable verbosity but I forgot. If not, we can just remove all. If yes, we could add some of those commented logs to that.

Reply:
We don’t have a logging mechanism, unfortunately